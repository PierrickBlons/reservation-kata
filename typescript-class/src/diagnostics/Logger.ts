interface Logger {
  error: (message: string) => void
  debug: (message: string) => void
  info: (message: string) => void
}

export default Logger
