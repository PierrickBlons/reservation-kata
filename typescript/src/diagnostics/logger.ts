type Logger = {
  error: (log: string) => void
  debug: (log: string) => void
  info: (log: string) => void
}

type LogLevel = 'info' | 'debug' | 'error'

export const spyLogger = (
  logsBucket: Array<{ level: LogLevel; log: string }>,
) =>
  ({
    error: (log: string) => logsBucket.push({ level: 'error', log }),
    debug: (log: string) => logsBucket.push({ level: 'debug', log }),
    info: (log: string) => logsBucket.push({ level: 'info', log }),
  }) satisfies Logger

export default Logger
