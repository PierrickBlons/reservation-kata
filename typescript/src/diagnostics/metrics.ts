type Metrics = {
  increment: (metricName: string) => void
}

export const metrics = {
  increment: (metricName: string) => {
    console.log(`Metric incremented: ${metricName}`)
  },
}

export const spyMetrics = (
  metricsBucket: Array<{ name: string; value: number }>,
) =>
  ({
    increment: (metricName: string) => {
      metricsBucket.push({ name: metricName, value: 1 }) // simple here has we won't have test making multiple reservations on the same hotel
    },
  }) satisfies Metrics

export default Metrics
