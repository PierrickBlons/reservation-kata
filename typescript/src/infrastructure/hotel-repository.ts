import Logger from '../diagnostics/logger'
import {
  Hotel,
  HotelName,
  RegisteredHotel,
  UnregisteredHotel,
} from '../domain/hotel'

export type HotelRepository = (logger: Logger) => {
  isRegistered: (hotelName: HotelName) => Hotel
}

export const inMemoryHotelRepository: HotelRepository = (logger) => ({
  isRegistered: (hotelName: HotelName) => {
    if (hotelName === 'Unknown Hotel') {
      logger.error("Hotel not found, can't proceed with reservation")
      return {
        type: 'unregisteredHotel',
        hotel: hotelName,
      } satisfies UnregisteredHotel
    }

    return {
      type: 'registeredHotel',
      hotel: hotelName,
      rooms: [
        {
          type: 'double',
        },
      ],
    } satisfies RegisteredHotel
  },
})
