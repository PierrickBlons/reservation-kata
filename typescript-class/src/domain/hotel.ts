import { ConfirmedRegistration, DraftReservation } from './reservation'

// Value types
export type HotelName = string & { __brand: 'hotelName' }
export type PaxNumber = number & { __brand: 'paxNumber' }
export type Reference = string & { __brand: 'reference' }

export class Stay {
  constructor(
    public readonly begin: Date,
    public readonly end: Date,
  ) {
    if (begin >= end) {
      throw new Error('End date must be after begin date')
    }
  }
}

// Hotel
export type UnregisteredHotel = { type: 'unregisteredHotel'; hotel: HotelName }

type Room = {
  type: 'double' | 'single'
  pax: PaxNumber
}

type RoomAvailability = {
  rooms: Room[]
  availabilityPeriod: Stay
}

export type RegisteredHotel = {
  type: 'registeredHotel'
  hotel: HotelName
  rooms: RoomAvailability[]
}

type HasAvailableRoom = (
  registeredHotel: RegisteredHotel,
  requestedStay: Stay,
) => RegisteredHotelRoom

const hasAvailableRoom: HasAvailableRoom = (
  registeredHotel: RegisteredHotel,
  requestedStay: Stay,
) => {
  if (registeredHotel.rooms.length === 0) {
    return {
      type: 'registeredHotelRoomNotAvailable',
      hotel: registeredHotel.hotel,
    } satisfies RegisteredHotelRoomNotAvailable
  }

  const roomAvailability = registeredHotel.rooms.find(
    (roomAvailability) =>
      roomAvailability.availabilityPeriod.begin <= requestedStay.begin &&
      roomAvailability.availabilityPeriod.end >= requestedStay.end,
  )

  if (roomAvailability) {
    return {
      type: 'registeredHotelRoomAvailable',
      hotel: registeredHotel.hotel,
      stayPeriod: requestedStay,
      room: roomAvailability.rooms[0],
    } satisfies RegisteredHotelRoomAvailable
  }

  return {
    type: 'registeredHotelRoomNotAvailable',
    hotel: registeredHotel.hotel,
  } satisfies RegisteredHotelRoomNotAvailable
}

export const RegisteredHotel = {
  hasAvailableRoom,
}

export type Hotel = RegisteredHotel | UnregisteredHotel

//HotelRoom
type RegisteredHotelRoomAvailable = {
  type: 'registeredHotelRoomAvailable'
  hotel: HotelName
  stayPeriod: Stay
  room: Room
}

type HasRoomForOccupancy = (
  hotel: RegisteredHotelRoomAvailable,
  pax: PaxNumber,
) => RegisteredHotelRoomOccupancy

type RegisteredHotelRoomNotAvailable = {
  type: 'registeredHotelRoomNotAvailable'
  hotel: HotelName
}

type RegisteredHotelRoom =
  | RegisteredHotelRoomAvailable
  | RegisteredHotelRoomNotAvailable

// Hotel Room Occupancy
const hasRoomForOccupancy: HasRoomForOccupancy = (
  registeredHotelWithAvailableRoom,
  pax,
) => {
  if (registeredHotelWithAvailableRoom.room.pax >= pax) {
    return RegisteredHotelRoomOccupancyAvailable.create(
      registeredHotelWithAvailableRoom,
    )
  }
  return RegisteredHotelRoomOccupancyInsufficient.create(
    registeredHotelWithAvailableRoom,
  )
}

export const RegisteredHotelRoomAvailable = {
  hasRoomForOccupancy,
}

type RegisteredHotelRoomOccupancyAvailable = {
  type: 'registeredHotelRoomOccupancyAvailable'
  hotel: HotelName
  stayPeriod: Stay
  room: Room
}

const createRegisteredHotelRoomOccupancyAvailable = ({
  hotel,
  stayPeriod,
  room,
}) =>
  ({
    type: 'registeredHotelRoomOccupancyAvailable',
    hotel,
    stayPeriod,
    room,
  }) satisfies RegisteredHotelRoomOccupancyAvailable

type InitializeReservation = (
  registeredHotelRoomOccupancyAvailable: RegisteredHotelRoomOccupancyAvailable,
) => DraftReservation

const initializeReservation: InitializeReservation = (
  registeredHotelRoomOccupancyAvailable,
) => {
  return {
    pax: registeredHotelRoomOccupancyAvailable.room.pax,
    stay: registeredHotelRoomOccupancyAvailable.stayPeriod,
  } satisfies DraftReservation
}

type ConfirmReservation = (
  RegisteredHotelRoomOccupancyAvailable: RegisteredHotelRoomOccupancyAvailable,
  draftReservation: DraftReservation,
) => ConfirmedRegistration

const confirmReservation: ConfirmReservation = (
  registeredHotelRoomOccupancyAvailable,
  draftReservation,
) => {
  return {
    hotel: registeredHotelRoomOccupancyAvailable.hotel,
    pax: draftReservation.pax,
    stay: draftReservation.stay,
    reference: 'GHRKJIK-45' as Reference,
  } satisfies ConfirmedRegistration
}

export const RegisteredHotelRoomOccupancyAvailable = {
  create: createRegisteredHotelRoomOccupancyAvailable,
  initializeReservation,
  confirmReservation,
}

type RegisteredHotelRoomOccupancyInsufficient = {
  type: 'registeredHotelRoomOccupancyInsufficient'
  hotel: HotelName
}

const createRegisteredHotelRoomOccupancyInsufficient = (
  registeredHotelWithAvailableRoom: RegisteredHotelRoomAvailable,
): RegisteredHotelRoomOccupancyInsufficient => {
  return {
    type: 'registeredHotelRoomOccupancyInsufficient',
    hotel: registeredHotelWithAvailableRoom.hotel,
  } satisfies RegisteredHotelRoomOccupancyInsufficient
}

export const RegisteredHotelRoomOccupancyInsufficient = {
  create: createRegisteredHotelRoomOccupancyInsufficient,
}

type RegisteredHotelRoomOccupancy =
  | RegisteredHotelRoomOccupancyAvailable
  | RegisteredHotelRoomOccupancyInsufficient
