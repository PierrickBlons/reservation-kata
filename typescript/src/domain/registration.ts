import { HotelName, PaxNumber, Reference } from './hotel'

export type ConfirmedRegistration = {
  hotel: HotelName
  pax: PaxNumber
  stay: { begin: Date; end: Date }
  reference: Reference
}
