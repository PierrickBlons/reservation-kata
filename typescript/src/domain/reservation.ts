import { HotelName, PaxNumber, Reference } from './hotel'

export type DraftReservation = {
  pax: PaxNumber
  stay: { begin: Date; end: Date }
}

export type ConfirmedRegistration = {
  hotel: HotelName
  pax: PaxNumber
  stay: { begin: Date; end: Date }
  reference: Reference
}
