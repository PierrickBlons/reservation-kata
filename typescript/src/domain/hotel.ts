import { ConfirmedRegistration, DraftReservation } from './reservation'

export type HotelName = string & { __brand: 'hotelName' }
export type PaxNumber = number & { __brand: 'paxNumber' }
export type Reference = string & { __brand: 'reference' }

export type UnregisteredHotel = { type: 'unregisteredHotel'; hotel: HotelName }

type Room = {
  type: 'double' | 'single'
  pax: PaxNumber
}

type RoomAvailability = {
  rooms: Room[]
  availabilityPeriod: Stay
}

export type RegisteredHotel = {
  type: 'registeredHotel'
  hotel: HotelName
  rooms: RoomAvailability[]
}

export type Hotel = RegisteredHotel | UnregisteredHotel

type HasAvailableRoom = (
  registeredHotel: RegisteredHotel,
  requestedStay: Stay,
) => RegisteredHotelRoom

type RegisteredHotelRoomAvailable = {
  type: 'registeredHotelRoomAvailable'
  hotel: HotelName
  stayPeriod: Stay
  room: Room
}

type RegisteredHotelRoomNotAvailable = {
  type: 'registeredHotelRoomNotAvailable'
  hotel: HotelName
}

type RegisteredHotelRoom =
  | RegisteredHotelRoomAvailable
  | RegisteredHotelRoomNotAvailable

type RegisteredHotelRoomOccupancyAvailable = {
  type: 'registeredHotelRoomOccupancyAvailable'
  hotel: HotelName
  stayPeriod: Stay
  room: Room
}

type RegisteredHotelRoomOccupancyInsufficient = {
  type: 'registeredHotelRoomOccupancyInsufficient'
  hotel: HotelName
}

type RegisteredHotelOccupancy =
  | RegisteredHotelRoomOccupancyAvailable
  | RegisteredHotelRoomOccupancyInsufficient

type HasRoomForOccupancy = (
  hotel: RegisteredHotelRoomAvailable,
  pax: PaxNumber,
) => RegisteredHotelOccupancy

export type Stay = {
  begin: Date
  end: Date
}

type InitializeReservation = (
  registeredHotelRoomOccupancyAvailable: RegisteredHotelRoomOccupancyAvailable,
) => DraftReservation

type ConfirmReservation = (
  RegisteredHotelRoomOccupancyAvailable: RegisteredHotelRoomOccupancyAvailable,
  draftReservation: DraftReservation,
) => ConfirmedRegistration

export const hasRoomForOccupancy: HasRoomForOccupancy = (
  registeredHotelWithAvailableRoom,
  pax,
) => {
  if (registeredHotelWithAvailableRoom.room.pax >= pax) {
    return {
      type: 'registeredHotelRoomOccupancyAvailable',
      hotel: registeredHotelWithAvailableRoom.hotel,
      stayPeriod: registeredHotelWithAvailableRoom.stayPeriod,
      room: registeredHotelWithAvailableRoom.room,
    } satisfies RegisteredHotelRoomOccupancyAvailable
  }
  return {
    type: 'registeredHotelRoomOccupancyInsufficient',
    hotel: registeredHotelWithAvailableRoom.hotel,
  } satisfies RegisteredHotelRoomOccupancyInsufficient
}

export const hasAvailableRoom: HasAvailableRoom = (
  registeredHotel: RegisteredHotel,
  requestedStay: Stay,
) => {
  if (registeredHotel.rooms.length === 0) {
    return {
      type: 'registeredHotelRoomNotAvailable',
      hotel: registeredHotel.hotel,
    } satisfies RegisteredHotelRoomNotAvailable
  }

  const roomAvailability = registeredHotel.rooms.find(
    (roomAvailability) =>
      roomAvailability.availabilityPeriod.begin <= requestedStay.begin &&
      roomAvailability.availabilityPeriod.end >= requestedStay.end,
  )

  if (roomAvailability) {
    return {
      type: 'registeredHotelRoomAvailable',
      hotel: registeredHotel.hotel,
      stayPeriod: requestedStay,
      room: roomAvailability.rooms[0],
    } satisfies RegisteredHotelRoomAvailable
  }
  return {
    type: 'registeredHotelRoomNotAvailable',
    hotel: registeredHotel.hotel,
  } satisfies RegisteredHotelRoomNotAvailable
}

export const initializeReservation: InitializeReservation = (
  registeredHotelRoomOccupancyAvailable,
) => {
  return {
    pax: registeredHotelRoomOccupancyAvailable.room.pax,
    stay: registeredHotelRoomOccupancyAvailable.stayPeriod,
  } satisfies DraftReservation
}

export const confirmReservation: ConfirmReservation = (
  registeredHotelRoomOccupancyAvailable,
  draftReservation,
) => {
  return {
    hotel: registeredHotelRoomOccupancyAvailable.hotel,
    pax: draftReservation.pax,
    stay: draftReservation.stay,
    reference: 'GHRKJIK-45' as Reference,
  } satisfies ConfirmedRegistration
}
