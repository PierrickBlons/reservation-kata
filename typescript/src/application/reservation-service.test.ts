import { describe, expect, it } from 'vitest'
import { metrics } from '../diagnostics/metrics'
import { HotelName, PaxNumber, Reference, Stay } from '../domain/hotel'
import { ConfirmedRegistration } from '../domain/reservation'
import { inMemoryHotelRepository } from '../infrastructure/hotel-repository'
import reservationService from './reservation-service'

const addDays = (date: Date, days: number): Date => {
  date.setDate(date.getDate() + days)
  return date
}

describe('reservation service make should', () => {
  it('give confirmedregistration with reference number', () => {
    const hotel = 'La Corniche' as HotelName
    const paxNumber = 2 as PaxNumber
    const begin = addDays(new Date(), 1)
    const end = addDays(new Date(), 2)

    const confirmedReservation = reservationService(
      console,
      metrics,
      inMemoryHotelRepository,
    ).make(hotel, paxNumber, { begin, end })

    expect(confirmedReservation).toEqual({
      hotel,
      pax: paxNumber,
      stay: {
        begin,
        end,
      } satisfies Stay,
      reference: 'GHRKJIK-45' as Reference,
    } satisfies ConfirmedRegistration)
  })

  it('give hotel not registered in our reservation system error message when hotel is not registered', () => {
    const unknownHotel = 'Unknown Hotel' as HotelName

    const failedReservation = reservationService(
      console,
      metrics,
      inMemoryHotelRepository,
    ).make(unknownHotel, 3 as PaxNumber, {
      begin: addDays(new Date(), 3),
      end: addDays(new Date(), 6),
    } satisfies Stay)

    expect(failedReservation).toEqual({
      Error: `Hotel ${unknownHotel} is not registered in our reservation system`,
    })
  })

  it('give roomNotAvailable when hotel has no room for the requested period', () => {
    const hotel = 'Sofitel' as HotelName

    const failedReservation = reservationService(
      console,
      metrics,
      inMemoryHotelRepository,
    ).make(hotel, 3 as PaxNumber, {
      begin: addDays(new Date(), 60),
      end: addDays(new Date(), 62),
    } satisfies Stay)

    expect(failedReservation).toEqual({
      Error: `Hotel ${hotel} doesn't have room for the selected period`,
    })
  })

  it('give occupancyNotAvailable when hotel has no occupancy for the period', () => {
    const unavailableHotel = 'Continental' as HotelName

    const requestedPaxNumber = 42 as PaxNumber
    const failedReservation = reservationService(
      console,
      metrics,
      inMemoryHotelRepository,
    ).make(unavailableHotel, requestedPaxNumber, {
      begin: addDays(new Date(), 7),
      end: addDays(new Date(), 9),
    } satisfies Stay)

    expect(failedReservation).toEqual({
      Error: `No room found for ${requestedPaxNumber} people`,
    })
  })
})
