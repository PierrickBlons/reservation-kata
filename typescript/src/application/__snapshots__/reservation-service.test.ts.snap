// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`reservation service make should > give confirmedregistration with reference number 1`] = `
{
  "logs": [
    {
      "level": "debug",
      "log": "Start making reservation",
    },
    {
      "level": "debug",
      "log": "Hotel La Corniche has room available",
    },
    {
      "level": "debug",
      "log": "Hotel La Corniche: Room found for 2",
    },
    {
      "level": "debug",
      "log": "Hotel La Corniche: Reservation option from 6/20/2025 to 6/21/2025 for 2 people",
    },
  ],
  "metrics": [
    {
      "name": "La Corniche.reservations",
      "value": 1,
    },
  ],
}
`;

exports[`reservation service make should > give error when requested stay endDate is before startDate 1`] = `
{
  "logs": [],
  "metrics": [],
}
`;

exports[`reservation service make should > give hotel not registered in our reservation system error message when hotel is not registered 1`] = `
{
  "logs": [
    {
      "level": "debug",
      "log": "Start making reservation",
    },
    {
      "level": "error",
      "log": "Hotel not found, can't proceed with reservation",
    },
    {
      "level": "info",
      "log": "Hotel Unknown Hotel is not registered in our reservation system",
    },
  ],
  "metrics": [
    {
      "name": "Unknown Hotel.reservations.hotelNotFound",
      "value": 1,
    },
  ],
}
`;

exports[`reservation service make should > give occupancyNotAvailable when hotel has no occupancy for the period 1`] = `
{
  "logs": [
    {
      "level": "debug",
      "log": "Start making reservation",
    },
    {
      "level": "debug",
      "log": "Hotel Continental has room available",
    },
    {
      "level": "info",
      "log": "No room found for 42 people",
    },
  ],
  "metrics": [
    {
      "name": "Continental.reservations.occupancyNotAvailable",
      "value": 1,
    },
  ],
}
`;

exports[`reservation service make should > give roomNotAvailable when hotel has no room for the requested period 1`] = `
{
  "logs": [
    {
      "level": "debug",
      "log": "Start making reservation",
    },
    {
      "level": "info",
      "log": "Hotel Sofitel doesn't have room for the selected period",
    },
  ],
  "metrics": [
    {
      "name": "Sofitel.reservations.roomNotAvailable",
      "value": 1,
    },
  ],
}
`;

exports[`reservation service make should > observability > should record a metric when a reservation is made 1`] = `
{
  "logs": [
    {
      "level": "debug",
      "log": "Start making reservation",
    },
    {
      "level": "debug",
      "log": "Hotel La Corniche has room available",
    },
    {
      "level": "debug",
      "log": "Hotel La Corniche: Room found for 2",
    },
    {
      "level": "debug",
      "log": "Hotel La Corniche: Reservation option from 6/20/2025 to 6/21/2025 for 2 people",
    },
  ],
  "metrics": [
    {
      "name": "La Corniche.reservations",
      "value": 1,
    },
  ],
}
`;
